<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>env.MRmega • env.MRmega</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="env.MRmega">
<meta property="og:description" content="env.MRmega">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">env.MRmega</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/envMRmega.html">env.MRmega</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/SiruRooney/env.MRmega/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>env.MRmega</h1>
                        <h4 data-toc-skip class="author">Siru Wang</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:sw2077@cam.ac.uk" class="email">sw2077@cam.ac.uk</a>
      
                  
            <h4 data-toc-skip class="date">17 May, 2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/SiruRooney/env.MRmega/blob/HEAD/vignettes/envMRmega.Rmd" class="external-link"><code>vignettes/envMRmega.Rmd</code></a></small>
      <div class="hidden name"><code>envMRmega.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><strong>env.MRmega</strong> (Environment-adjusted MR-MEGA) is a
package to identify genetic variants that are associated with each
trait, across cohorts, while adjusting for differing environmental
exposures between cohorts. For the environment-adjusted MR-MEGA
approach, we will adapt the original MR-MEGA meta-regression framework
to account for environmental exposures that are specific each cohort.
Apart from environment-adjusted MR-MEGA methodology, the R package also
includes pre-processing of the inputted GWAS files, calculation of axes
of genetic variation and the original MR-MEGA methodology.</p>
<p>This vignette introduces the whole meta-analysis procedure using the
original MR-MEGA approach and the environment-adjusted MR-MEGA
approach.</p>
</div>
<div class="section level2">
<h2 id="gwas-data-format">GWAS data format<a class="anchor" aria-label="anchor" href="#gwas-data-format"></a>
</h2>
<p>No matter which meta-analysis methods (environment-adjusted MR-MEGA
or MR-MEGA) are applied, a list of GWAS datasets would be the main input
as required, and each GWAS data must contain mandatory columns. Details
of GWAS format would be given by a simple simulation.</p>
<p>For convenience of illustration, we would grossly simulate several
GWAS datasets containing medium-sized genetic variants and the
associated genetic information. In the simulated data, we considered 7
sub-populations of African ancestry which were collected from Phase 3 of
the 1000 Genomes Project<span class="citation">(<a href="#ref-10002015global">Consortium et al. 2015</a>)</span>,see <a href="https://ctg.cncr.nl/software/MAGMA/ref_data/" class="external-link uri">https://ctg.cncr.nl/software/MAGMA/ref_data/</a>. The 7
populations were coded as <strong>ACB</strong>, <strong>ASW</strong>,
<strong>ESN</strong>, <strong>GWD</strong>, <strong>LWK</strong>,
<strong>MSL</strong>, <strong>YRI</strong>. Among the 7 GWAS raw data,
we randomly selected 50 genetic variants across 23 chromosomes to create
7 medium-sized GWAS files. The format of each file has mandatory column
headers, shown as <span style="color: #0000FF;">MR-MEGA</span> software
(<a href="https://genomics.ut.ee/en/tools" class="external-link uri">https://genomics.ut.ee/en/tools</a>):</p>
<ul>
<li>
<code>MARKERNAME</code> - snp name</li>
<li>
<code>N</code> - sample size</li>
<li>
<code>EAF</code> - effect allele frequency</li>
<li>
<code>EA</code> - effect allele</li>
<li>
<code>NEA</code> - non effect allele</li>
<li>
<code>CHROMOSOME</code> - chromosome of marker</li>
<li>
<code>POSITION</code> - position of marker</li>
<li>
<code>BETA</code> - beta</li>
<li>
<code>SE</code> - std.error</li>
</ul>
<p>When simulating quantitative traits, in the specified African
populations (<strong>ESN</strong>, <strong>GWD</strong>,
<strong>LWK</strong>), we consider equal-sized populations (4K in each
population). Then we set three genetic
variants,<code>rs201029612</code>, <code>rs79862059</code> and
<code>rs77833422</code>, which are associated with the quantitative
traits. To mimic the real GWAS data in which some genetic variants would
not exist in all populations, we made that the two genetic variants,
<code>rs190897515</code> and <code>rs577430535</code>, exist in
<strong>ACB</strong> population only; one genetic variant
<code>rs185581532</code> exists in 5 populations only and one genetic
variant <code>rs572565163</code> exists in 4 populations only. The
details of 7 simulated GWAS data are saved in <code>afr.gwas</code>
object as a list.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://github.com/SiruRooney/env.MRmega" class="external-link">env.MRmega</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: env.MRmega</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">afr.gwas</span>,<span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">x</span>,<span class="fl">2</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span>
<span><span class="co">#&gt; $acb</span></span>
<span><span class="co">#&gt;     MARKERNAME     N    EAF      BETA       SE CHROMOSOME POSITION     EA</span></span>
<span><span class="co">#&gt;         &lt;char&gt; &lt;int&gt;  &lt;num&gt;     &lt;num&gt;    &lt;num&gt;      &lt;int&gt;    &lt;int&gt; &lt;char&gt;</span></span>
<span><span class="co">#&gt; 1:  rs10429359  4000 0.3646 -0.000422 0.010527          8 72918699      C</span></span>
<span><span class="co">#&gt; 2: rs115995911  4000 0.1042 -0.001478 0.016606         10 73624098      T</span></span>
<span><span class="co">#&gt;       NEA</span></span>
<span><span class="co">#&gt;    &lt;char&gt;</span></span>
<span><span class="co">#&gt; 1:      G</span></span>
<span><span class="co">#&gt; 2:      C</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $asw</span></span>
<span><span class="co">#&gt;     MARKERNAME     N    EAF      BETA       SE CHROMOSOME POSITION     EA</span></span>
<span><span class="co">#&gt;         &lt;char&gt; &lt;int&gt;  &lt;num&gt;     &lt;num&gt;    &lt;num&gt;      &lt;int&gt;    &lt;int&gt; &lt;char&gt;</span></span>
<span><span class="co">#&gt; 1:  rs10429359  4000 0.3115 -0.018387 0.010747          8 72918699      C</span></span>
<span><span class="co">#&gt; 2: rs115995911  4000 0.1885 -0.005697 0.012782         10 73624098      T</span></span>
<span><span class="co">#&gt;       NEA</span></span>
<span><span class="co">#&gt;    &lt;char&gt;</span></span>
<span><span class="co">#&gt; 1:      G</span></span>
<span><span class="co">#&gt; 2:      C</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $esn</span></span>
<span><span class="co">#&gt;     MARKERNAME     N    EAF      BETA       SE CHROMOSOME POSITION     EA</span></span>
<span><span class="co">#&gt;         &lt;char&gt; &lt;int&gt;  &lt;num&gt;     &lt;num&gt;    &lt;num&gt;      &lt;int&gt;    &lt;int&gt; &lt;char&gt;</span></span>
<span><span class="co">#&gt; 1:  rs10429359  4000 0.3283  0.000632 0.010673          8 72918699      C</span></span>
<span><span class="co">#&gt; 2: rs115995911  4000 0.1515 -0.023159 0.014028         10 73624098      T</span></span>
<span><span class="co">#&gt;       NEA</span></span>
<span><span class="co">#&gt;    &lt;char&gt;</span></span>
<span><span class="co">#&gt; 1:      G</span></span>
<span><span class="co">#&gt; 2:      C</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $gwd</span></span>
<span><span class="co">#&gt;     MARKERNAME     N    EAF      BETA       SE CHROMOSOME POSITION     EA</span></span>
<span><span class="co">#&gt;         &lt;char&gt; &lt;int&gt;  &lt;num&gt;     &lt;num&gt;    &lt;num&gt;      &lt;int&gt;    &lt;int&gt; &lt;char&gt;</span></span>
<span><span class="co">#&gt; 1:  rs10429359  4000 0.3805  0.012997 0.010300          8 72918699      C</span></span>
<span><span class="co">#&gt; 2: rs115995911  4000 0.2212 -0.026722 0.012143         10 73624098      T</span></span>
<span><span class="co">#&gt;       NEA</span></span>
<span><span class="co">#&gt;    &lt;char&gt;</span></span>
<span><span class="co">#&gt; 1:      G</span></span>
<span><span class="co">#&gt; 2:      C</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $lwk</span></span>
<span><span class="co">#&gt;     MARKERNAME     N     EAF     BETA       SE CHROMOSOME POSITION     EA</span></span>
<span><span class="co">#&gt;         &lt;char&gt; &lt;int&gt;   &lt;num&gt;    &lt;num&gt;    &lt;num&gt;      &lt;int&gt;    &lt;int&gt; &lt;char&gt;</span></span>
<span><span class="co">#&gt; 1:  rs10429359  4000 0.30810 0.021422 0.010870          8 72918699      C</span></span>
<span><span class="co">#&gt; 2: rs115995911  4000 0.07071 0.000319 0.020456         10 73624098      T</span></span>
<span><span class="co">#&gt;       NEA</span></span>
<span><span class="co">#&gt;    &lt;char&gt;</span></span>
<span><span class="co">#&gt; 1:      G</span></span>
<span><span class="co">#&gt; 2:      C</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $msl</span></span>
<span><span class="co">#&gt;     MARKERNAME     N    EAF      BETA       SE CHROMOSOME POSITION     EA</span></span>
<span><span class="co">#&gt;         &lt;char&gt; &lt;int&gt;  &lt;num&gt;     &lt;num&gt;    &lt;num&gt;      &lt;int&gt;    &lt;int&gt; &lt;char&gt;</span></span>
<span><span class="co">#&gt; 1:  rs10429359  4000 0.3118 -0.011872 0.010834          8 72918699      C</span></span>
<span><span class="co">#&gt; 2: rs115995911  4000 0.1647  0.015265 0.013528         10 73624098      T</span></span>
<span><span class="co">#&gt;       NEA</span></span>
<span><span class="co">#&gt;    &lt;char&gt;</span></span>
<span><span class="co">#&gt; 1:      G</span></span>
<span><span class="co">#&gt; 2:      C</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $yri</span></span>
<span><span class="co">#&gt;     MARKERNAME     N    EAF      BETA       SE CHROMOSOME POSITION     EA</span></span>
<span><span class="co">#&gt;         &lt;char&gt; &lt;int&gt;  &lt;num&gt;     &lt;num&gt;    &lt;num&gt;      &lt;int&gt;    &lt;int&gt; &lt;char&gt;</span></span>
<span><span class="co">#&gt; 1:  rs10429359  4000 0.3935  0.016963 0.010342          8 72918699      C</span></span>
<span><span class="co">#&gt; 2: rs115995911  4000 0.1898 -0.009583 0.012878         10 73624098      T</span></span>
<span><span class="co">#&gt;       NEA</span></span>
<span><span class="co">#&gt;    &lt;char&gt;</span></span>
<span><span class="co">#&gt; 1:      G</span></span>
<span><span class="co">#&gt; 2:      C</span></span></code></pre></div>
<p>In real life, GWAS statistic data of each cohort would be saved as a
file (“.txt”), even a compressed file (“.txt.gz”), containing tens of
thousands of genetic variants information. We could create a input file
which contains all study files and use <code>readInputfile</code>
function to aggregate several GWAS files into a list.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_name</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"~/Desktop/mr-mega-r/MR_MEGA_ENV_git/g1000_afr_7gwas_git/afr_acb_EGL20.txt"</span>,</span>
<span>            <span class="st">"~/Desktop/mr-mega-r/MR_MEGA_ENV_git/g1000_afr_7gwas_git/afr_asw_EGL20.txt"</span>,</span>
<span>            <span class="st">"~/Desktop/mr-mega-r/MR_MEGA_ENV_git/g1000_afr_7gwas_git/afr_esn_EGL20.txt"</span>,</span>
<span>            <span class="st">"~/Desktop/mr-mega-r/MR_MEGA_ENV_git/g1000_afr_7gwas_git/afr_gwd_EGL20.txt"</span>,</span>
<span>            <span class="st">"~/Desktop/mr-mega-r/MR_MEGA_ENV_git/g1000_afr_7gwas_git/afr_lwk_EGL20.txt"</span>,</span>
<span>            <span class="st">"~/Desktop/mr-mega-r/MR_MEGA_ENV_git/g1000_afr_7gwas_git/afr_msl_EGL20.txt"</span>,</span>
<span>            <span class="st">"~/Desktop/mr-mega-r/MR_MEGA_ENV_git/g1000_afr_7gwas_git/afr_yri_EGL20.txt"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">out</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/readInputfile.html">readInputfile</a></span><span class="op">(</span><span class="va">data_name</span>,qt<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "i in filesize 1"</span></span>
<span><span class="co">#&gt; [1] "Reading file ~/Desktop/mr-mega-r/MR_MEGA_ENV_git/g1000_afr_7gwas_git/afr_acb_EGL20.txt"</span></span>
<span><span class="co">#&gt; [1] "i in filesize 2"</span></span>
<span><span class="co">#&gt; [1] "Reading file ~/Desktop/mr-mega-r/MR_MEGA_ENV_git/g1000_afr_7gwas_git/afr_asw_EGL20.txt"</span></span>
<span><span class="co">#&gt; [1] "i in filesize 3"</span></span>
<span><span class="co">#&gt; [1] "Reading file ~/Desktop/mr-mega-r/MR_MEGA_ENV_git/g1000_afr_7gwas_git/afr_esn_EGL20.txt"</span></span>
<span><span class="co">#&gt; [1] "i in filesize 4"</span></span>
<span><span class="co">#&gt; [1] "Reading file ~/Desktop/mr-mega-r/MR_MEGA_ENV_git/g1000_afr_7gwas_git/afr_gwd_EGL20.txt"</span></span>
<span><span class="co">#&gt; [1] "i in filesize 5"</span></span>
<span><span class="co">#&gt; [1] "Reading file ~/Desktop/mr-mega-r/MR_MEGA_ENV_git/g1000_afr_7gwas_git/afr_lwk_EGL20.txt"</span></span>
<span><span class="co">#&gt; [1] "i in filesize 6"</span></span>
<span><span class="co">#&gt; [1] "Reading file ~/Desktop/mr-mega-r/MR_MEGA_ENV_git/g1000_afr_7gwas_git/afr_msl_EGL20.txt"</span></span>
<span><span class="co">#&gt; [1] "i in filesize 7"</span></span>
<span><span class="co">#&gt; [1] "Reading file ~/Desktop/mr-mega-r/MR_MEGA_ENV_git/g1000_afr_7gwas_git/afr_yri_EGL20.txt"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">out</span>,<span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">x</span>,<span class="fl">2</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span>
<span><span class="co">#&gt; $Pop1</span></span>
<span><span class="co">#&gt;     MARKERNAME CHROMOSOME POSITION     EA    NEA    EAF      BETA       SE</span></span>
<span><span class="co">#&gt;         &lt;char&gt;      &lt;int&gt;    &lt;int&gt; &lt;char&gt; &lt;char&gt;  &lt;num&gt;     &lt;num&gt;    &lt;num&gt;</span></span>
<span><span class="co">#&gt; 1:  rs10429359          8 72918699      C      G 0.3646 -0.000422 0.010527</span></span>
<span><span class="co">#&gt; 2: rs115995911         10 73624098      T      C 0.1042 -0.001478 0.016606</span></span>
<span><span class="co">#&gt;        N</span></span>
<span><span class="co">#&gt;    &lt;int&gt;</span></span>
<span><span class="co">#&gt; 1:  4000</span></span>
<span><span class="co">#&gt; 2:  4000</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $Pop2</span></span>
<span><span class="co">#&gt;     MARKERNAME CHROMOSOME POSITION     EA    NEA    EAF      BETA       SE</span></span>
<span><span class="co">#&gt;         &lt;char&gt;      &lt;int&gt;    &lt;int&gt; &lt;char&gt; &lt;char&gt;  &lt;num&gt;     &lt;num&gt;    &lt;num&gt;</span></span>
<span><span class="co">#&gt; 1:  rs10429359          8 72918699      C      G 0.3115 -0.018387 0.010747</span></span>
<span><span class="co">#&gt; 2: rs115995911         10 73624098      T      C 0.1885 -0.005697 0.012782</span></span>
<span><span class="co">#&gt;        N</span></span>
<span><span class="co">#&gt;    &lt;int&gt;</span></span>
<span><span class="co">#&gt; 1:  4000</span></span>
<span><span class="co">#&gt; 2:  4000</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $Pop3</span></span>
<span><span class="co">#&gt;     MARKERNAME CHROMOSOME POSITION     EA    NEA    EAF      BETA       SE</span></span>
<span><span class="co">#&gt;         &lt;char&gt;      &lt;int&gt;    &lt;int&gt; &lt;char&gt; &lt;char&gt;  &lt;num&gt;     &lt;num&gt;    &lt;num&gt;</span></span>
<span><span class="co">#&gt; 1:  rs10429359          8 72918699      C      G 0.3283  0.000632 0.010673</span></span>
<span><span class="co">#&gt; 2: rs115995911         10 73624098      T      C 0.1515 -0.023159 0.014028</span></span>
<span><span class="co">#&gt;        N</span></span>
<span><span class="co">#&gt;    &lt;int&gt;</span></span>
<span><span class="co">#&gt; 1:  4000</span></span>
<span><span class="co">#&gt; 2:  4000</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $Pop4</span></span>
<span><span class="co">#&gt;     MARKERNAME CHROMOSOME POSITION     EA    NEA    EAF      BETA       SE</span></span>
<span><span class="co">#&gt;         &lt;char&gt;      &lt;int&gt;    &lt;int&gt; &lt;char&gt; &lt;char&gt;  &lt;num&gt;     &lt;num&gt;    &lt;num&gt;</span></span>
<span><span class="co">#&gt; 1:  rs10429359          8 72918699      C      G 0.3805  0.012997 0.010300</span></span>
<span><span class="co">#&gt; 2: rs115995911         10 73624098      T      C 0.2212 -0.026722 0.012143</span></span>
<span><span class="co">#&gt;        N</span></span>
<span><span class="co">#&gt;    &lt;int&gt;</span></span>
<span><span class="co">#&gt; 1:  4000</span></span>
<span><span class="co">#&gt; 2:  4000</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $Pop5</span></span>
<span><span class="co">#&gt;     MARKERNAME CHROMOSOME POSITION     EA    NEA     EAF     BETA       SE</span></span>
<span><span class="co">#&gt;         &lt;char&gt;      &lt;int&gt;    &lt;int&gt; &lt;char&gt; &lt;char&gt;   &lt;num&gt;    &lt;num&gt;    &lt;num&gt;</span></span>
<span><span class="co">#&gt; 1:  rs10429359          8 72918699      C      G 0.30810 0.021422 0.010870</span></span>
<span><span class="co">#&gt; 2: rs115995911         10 73624098      T      C 0.07071 0.000319 0.020456</span></span>
<span><span class="co">#&gt;        N</span></span>
<span><span class="co">#&gt;    &lt;int&gt;</span></span>
<span><span class="co">#&gt; 1:  4000</span></span>
<span><span class="co">#&gt; 2:  4000</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $Pop6</span></span>
<span><span class="co">#&gt;     MARKERNAME CHROMOSOME POSITION     EA    NEA    EAF      BETA       SE</span></span>
<span><span class="co">#&gt;         &lt;char&gt;      &lt;int&gt;    &lt;int&gt; &lt;char&gt; &lt;char&gt;  &lt;num&gt;     &lt;num&gt;    &lt;num&gt;</span></span>
<span><span class="co">#&gt; 1:  rs10429359          8 72918699      C      G 0.3118 -0.011872 0.010834</span></span>
<span><span class="co">#&gt; 2: rs115995911         10 73624098      T      C 0.1647  0.015265 0.013528</span></span>
<span><span class="co">#&gt;        N</span></span>
<span><span class="co">#&gt;    &lt;int&gt;</span></span>
<span><span class="co">#&gt; 1:  4000</span></span>
<span><span class="co">#&gt; 2:  4000</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $Pop7</span></span>
<span><span class="co">#&gt;     MARKERNAME CHROMOSOME POSITION     EA    NEA    EAF      BETA       SE</span></span>
<span><span class="co">#&gt;         &lt;char&gt;      &lt;int&gt;    &lt;int&gt; &lt;char&gt; &lt;char&gt;  &lt;num&gt;     &lt;num&gt;    &lt;num&gt;</span></span>
<span><span class="co">#&gt; 1:  rs10429359          8 72918699      C      G 0.3935  0.016963 0.010342</span></span>
<span><span class="co">#&gt; 2: rs115995911         10 73624098      T      C 0.1898 -0.009583 0.012878</span></span>
<span><span class="co">#&gt;        N</span></span>
<span><span class="co">#&gt;    &lt;int&gt;</span></span>
<span><span class="co">#&gt; 1:  4000</span></span>
<span><span class="co">#&gt; 2:  4000</span></span></code></pre></div>
<p><span style="color: #0000FF;">Note</span>:</p>
<ul>
<li>For convenience of simplicity, we do not take account of sex
stratification in this case. In real life, each GWAS file can be
stratified by sex into male and female cohorts. The 7 GWAS files then
would be divided into 14 cohorts, then the environment-adjusted MR-MEGA
and the original MR-MEGA approaches can also be used for sex-stratified
GWAS data.</li>
</ul>
<p>For the environment-adjusted meta-regression analysis, besides the
above mentioned mandatory columns in each GWAS study file, we also need
to add cohort-specific environmental exposure information. Here we
assume a binary environmental exposure, smoking status (1/0). Within
each cohort, smoking status=1 for individuals having smoking habits and
smoking status=0 for individuals having no smoking habits.</p>
<p><span style="color: #0000FF;">Note</span>:</p>
<ul>
<li>For 14 sex-stratified cohorts, the individual-level data for
environment covariates would comprise one binary smoking status and one
binary sex indicator (female=1/male=0). With each sex-stratified cohort,
sex indicator =1 for all female cohorts and sex indicator =0 for all
male cohorts.</li>
</ul>
</div>
<div class="section level2">
<h2 id="simulation-example">Simulation example<a class="anchor" aria-label="anchor" href="#simulation-example"></a>
</h2>
<div class="section level3">
<h3 id="packages-installation-and-loading">Packages installation and loading<a class="anchor" aria-label="anchor" href="#packages-installation-and-loading"></a>
</h3>
<p>First, we load the packages necessary for analysis, and check the
version of each package.The provided demo runs in less than 2 seconds
including preparation and execution of both MR-MEGA/environment-adjusted
MR-MEGA procedures.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#cran</span></span>
<span><span class="va">cran.pkgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'data.table'</span>, <span class="st">'doMC'</span>, <span class="st">'doParallel'</span>,<span class="st">'dplyr'</span>,<span class="st">'iterators'</span>,<span class="st">'parallel'</span>,<span class="st">'R.utils'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load packages </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">cran.pkgs</span>, <span class="va">require</span>, character.only <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="data-pre-processing">Data pre-processing<a class="anchor" aria-label="anchor" href="#data-pre-processing"></a>
</h3>
<p><strong>Calculate PCs</strong></p>
<p>Before fitting the environment-adjusted meta-regression model and the
original meta-regression model <span class="citation">(<a href="#ref-magi2017trans">Mägi et al. 2017</a>)</span>, we need to
specify the number of PCs (<span style="color: #0000FF;">T</span>),
default=2. Specifically, we used a subset of genetic variants with
MAF&gt;5% in all population files to derive Euclidean distances matrix.
Then the two axes of genetic variation were derived from the
multi-dimensional scaling of the distance matrix in the corresponding
reference panel.</p>
<p><span style="color: #0000FF;">Note</span>: the number of axes of
genetic variation and the number of environmental covariates would
depend on the diversity of cohorts, but the limitation of the number of
axes of genetic variation (<span style="color: #0000FF;">T</span>) and
the number of environmental covariates (<span style="color: #0000FF;">S</span>) should be less than the number of
cohorts (<span style="color: #0000FF;">K</span>) - 2.</p>
<p>The function used for calculating PCs would be built in the function
<code>MR_mega_prep</code> which is used for preparing inputted data in
the environment-adjusted meta-regression and the original
meta-regression models. For convenience of PC calculation independently,
we extracted the function used for calculating PCs
<code>MR_mega_MDS</code> out. Since the simulated GWAS data contain 50
genetic variants only, we would derive PCs from the subset of genetic
variants with MAF&gt;5% in the whole of the 7 GWAS raw files, each file
consisting of thousands of genetic variants across 23 chromosomes.</p>
<p>For running <code>MR_mega_MDS</code> function, you can create a
vector containing all population file names and the corresponding
locations. Then you need to point out the location of the output, named
by <code>batch_loc</code>.</p>
<p><span style="color: #0000FF;">Note</span>: The data are pre-loaded
and these lines should not be run in the demo.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_name</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"~/MR_MEGA/MR_MEGA_ENV_gitdta/g1000_afr_acb/g1000_afr_acb.inf"</span>,</span>
<span>            <span class="st">"~/MR_MEGA/MR_MEGA_ENV_gitdta/g1000_afr_asw/g1000_afr_asw.inf"</span>,</span>
<span>            <span class="st">"~/MR_MEGA/MR_MEGA_ENV_gitdta/g1000_afr_esn/g1000_afr_esn.inf"</span>,</span>
<span>            <span class="st">"~/MR_MEGA/MR_MEGA_ENV_gitdta/g1000_afr_gwd/g1000_afr_gwd.inf"</span>,</span>
<span>            <span class="st">"~/MR_MEGA/MR_MEGA_ENV_gitdta/g1000_afr_lwk/g1000_afr_lwk.inf"</span>,</span>
<span>            <span class="st">"~/MR_MEGA/MR_MEGA_ENV_gitdta/g1000_afr_msl/g1000_afr_msl.inf"</span>,</span>
<span>            <span class="st">"~/MR_MEGA/MR_MEGA_ENV_gitdta/g1000_afr_yri/g1000_afr_yri.inf"</span><span class="op">)</span></span>
<span><span class="va">batch_loc</span><span class="op">=</span><span class="st">"~/MR_MEGA/MR_MEGA_ENV_gitdta/"</span></span>
<span><span class="fu"><a href="../reference/MR_mega_MDS.html">MR_mega_MDS</a></span><span class="op">(</span><span class="va">data_name</span>,<span class="va">batch_loc</span>,qt<span class="op">=</span><span class="cn">TRUE</span>,pcCount<span class="op">=</span><span class="fl">2</span>,ncores<span class="op">=</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<p>In the above command, <code>MR_mega_MDS</code> function would output
two files, one file containing axes of genetic variants and another
containing multi-dimensional scaling (MDS) of Euclidean distances
matrix. After inputting 7 GWAS files, we get the PCs shown as
follows.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pcs</span></span>
<span><span class="co">#&gt;            pc1          pc2</span></span>
<span><span class="co">#&gt;          &lt;num&gt;        &lt;num&gt;</span></span>
<span><span class="co">#&gt; 1: -0.01143282  0.004831679</span></span>
<span><span class="co">#&gt; 2: -0.04685165  0.020099247</span></span>
<span><span class="co">#&gt; 3:  0.01282844 -0.012446795</span></span>
<span><span class="co">#&gt; 4:  0.02037034  0.020354682</span></span>
<span><span class="co">#&gt; 5: -0.01215912 -0.040216803</span></span>
<span><span class="co">#&gt; 6:  0.02405165  0.014331996</span></span>
<span><span class="co">#&gt; 7:  0.01319316 -0.006954006</span></span></code></pre></div>
<p><strong>Categorize BETA, SE and other genetic variant information
into different objects</strong></p>
<p>Based on the 7 simulated GWAS files, we categorized <code>BETA</code>
and <code>SE</code> into two objects which would be saved as a list via
<code>MR_mega_prep</code> function.</p>
<p><span style="color: #0000FF;">Note</span>:</p>
<ul>
<li><p>It is required that the order of names of <code>afr.gwas</code>
should always be consistent with the order of <code>data_name</code> as
input in <code>MR_mega_MDS</code> function.</p></li>
<li><p>Considering GWAS files might contain thousands of genetic
variants in real life, we can also divide <code>BETA</code> object and
<code>SE</code> object into several batches. When conducting the
environment-adjusted MR-MEGA approach and the original MR-MEGA approach,
we could input these batches into the associated functions and
performing them in parallel.</p></li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">################env_adjusted MR-MEGA#####################</span></span>
<span><span class="fu">tictoc</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tictoc/man/tic.html" class="external-link">tic</a></span><span class="op">(</span><span class="st">"Preparation for performing MR-MEGA/env-MR-MEGA approach"</span><span class="op">)</span></span>
<span><span class="va">gwas_input</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/MR_mega_prep.html">MR_mega_prep</a></span><span class="op">(</span>sum_stat_pop<span class="op">=</span><span class="va">afr.gwas</span>,n_cohort<span class="op">=</span><span class="fl">7</span>,usefor<span class="op">=</span><span class="st">"env-mr-mega"</span>,qt<span class="op">=</span><span class="cn">TRUE</span>,pcCount<span class="op">=</span><span class="fl">2</span>,calpc<span class="op">=</span><span class="cn">FALSE</span>,envCount<span class="op">=</span><span class="fl">1</span>,ncores<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Parallel running and use2 cores."</span></span>
<span><span class="co">#&gt; [1] "All GWAS are aligned to the same reference allele"</span></span>
<span><span class="co">#&gt; [1] "Set the ancestry 1 as the reference allele"</span></span>
<span><span class="co">#&gt; There are  50 different SNPs among 7 populations</span></span>
<span><span class="fu">tictoc</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tictoc/man/tic.html" class="external-link">toc</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Preparation for performing MR-MEGA/env-MR-MEGA approach: 0.335 sec elapsed</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gwas_input</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "beta_pop"          "invse2_pop"        "marker_inf_pop"   </span></span>
<span><span class="co">#&gt; [4] "cohort_count_filt" "sel_gene"</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gwas_input</span><span class="op">$</span><span class="va">beta_pop</span>,<span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt;    MARKERNAME      Pop1      Pop2      Pop3      Pop4     Pop5      Pop6</span></span>
<span><span class="co">#&gt; 1  rs10429359 -0.000422 -0.018387  0.000632  0.012997 0.021422 -0.011872</span></span>
<span><span class="co">#&gt; 2 rs115995911 -0.001478 -0.005697 -0.023159 -0.026722 0.000319  0.015265</span></span>
<span><span class="co">#&gt;        Pop7</span></span>
<span><span class="co">#&gt; 1  0.016963</span></span>
<span><span class="co">#&gt; 2 -0.009583</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gwas_input</span><span class="op">$</span><span class="va">invse2_pop</span>,<span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt;    MARKERNAME Pop1 Pop2 Pop3 Pop4 Pop5 Pop6 Pop7</span></span>
<span><span class="co">#&gt; 1  rs10429359 9024 8658 8779 9426 8463 8520 9350</span></span>
<span><span class="co">#&gt; 2 rs115995911 3626 6121 5082 6782 2390 5464 6030</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gwas_input</span><span class="op">$</span><span class="va">marker_inf_pop</span>,<span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt;    MARKERNAME CHROMOSOME EA NEA     N   EAF_avg      dir</span></span>
<span><span class="co">#&gt; 1  rs10429359          8  C   G 28000 0.3426143 ?--+++-+</span></span>
<span><span class="co">#&gt; 2 rs115995911         10  T   C 28000 0.1558014 ?----++-</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gwas_input</span><span class="op">$</span><span class="va">cohort_count_filt</span>,<span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt;    MARKERNAME Count_cohorts</span></span>
<span><span class="co">#&gt; 1  rs10429359             7</span></span>
<span><span class="co">#&gt; 2 rs115995911             7</span></span>
<span><span class="va">gwas_input</span><span class="op">$</span><span class="va">sel_gene</span></span>
<span><span class="co">#&gt;    MARKERNAME beta0 se0 beta1 se1 beta2 se2 beta3 se3 beta4 se4 beta5 se5 beta6</span></span>
<span><span class="co">#&gt; 1 rs190897515    NA  NA    NA  NA    NA  NA    NA  NA    NA  NA    NA  NA    NA</span></span>
<span><span class="co">#&gt; 2 rs577430535    NA  NA    NA  NA    NA  NA    NA  NA    NA  NA    NA  NA    NA</span></span>
<span><span class="co">#&gt; 3 rs185581532    NA  NA    NA  NA    NA  NA    NA  NA    NA  NA    NA  NA    NA</span></span>
<span><span class="co">#&gt; 4 rs572565163    NA  NA    NA  NA    NA  NA    NA  NA    NA  NA    NA  NA    NA</span></span>
<span><span class="co">#&gt;   se6 chisq_association ndf_association pvalue_association chisq_heter</span></span>
<span><span class="co">#&gt; 1  NA                NA              NA                 NA          NA</span></span>
<span><span class="co">#&gt; 2  NA                NA              NA                 NA          NA</span></span>
<span><span class="co">#&gt; 3  NA                NA              NA                 NA          NA</span></span>
<span><span class="co">#&gt; 4  NA                NA              NA                 NA          NA</span></span>
<span><span class="co">#&gt;   ndf_chisq_heter pvalue_heter chisq_residual ndf_residual pvalue_residual</span></span>
<span><span class="co">#&gt; 1              NA           NA             NA           NA              NA</span></span>
<span><span class="co">#&gt; 2              NA           NA             NA           NA              NA</span></span>
<span><span class="co">#&gt; 3              NA           NA             NA           NA              NA</span></span>
<span><span class="co">#&gt; 4              NA           NA             NA           NA              NA</span></span>
<span><span class="co">#&gt;           comments</span></span>
<span><span class="co">#&gt; 1 SmallcohortCount</span></span>
<span><span class="co">#&gt; 2 SmallcohortCount</span></span>
<span><span class="co">#&gt; 3 SmallcohortCount</span></span>
<span><span class="co">#&gt; 4 SmallcohortCount</span></span>
<span><span class="co">################MR-MEGA#####################</span></span>
<span><span class="va">gwas_input</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/MR_mega_prep.html">MR_mega_prep</a></span><span class="op">(</span>sum_stat_pop<span class="op">=</span><span class="va">afr.gwas</span>,n_cohort<span class="op">=</span><span class="fl">7</span>,usefor<span class="op">=</span><span class="st">"mr-mega"</span>,qt<span class="op">=</span><span class="cn">TRUE</span>,pcCount<span class="op">=</span><span class="fl">2</span>,calpc<span class="op">=</span><span class="cn">FALSE</span>,ncores<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Parallel running and use2 cores."</span></span>
<span><span class="co">#&gt; [1] "All GWAS are aligned to the same reference allele"</span></span>
<span><span class="co">#&gt; [1] "Set the ancestry 1 as the reference allele"</span></span>
<span><span class="co">#&gt; There are  50 different SNPs among 7 populations</span></span>
<span><span class="va">gwas_input</span><span class="op">$</span><span class="va">sel_gene</span></span>
<span><span class="co">#&gt;    MARKERNAME beta0 se0 beta1 se1 beta2 se2 beta3 se3 beta4 se4</span></span>
<span><span class="co">#&gt; 1 rs190897515    NA  NA    NA  NA    NA  NA    NA  NA    NA  NA</span></span>
<span><span class="co">#&gt; 2 rs577430535    NA  NA    NA  NA    NA  NA    NA  NA    NA  NA</span></span>
<span><span class="co">#&gt; 3 rs572565163    NA  NA    NA  NA    NA  NA    NA  NA    NA  NA</span></span>
<span><span class="co">#&gt;   chisq_association ndf_association pvalue_association chisq_heter</span></span>
<span><span class="co">#&gt; 1                NA              NA                 NA          NA</span></span>
<span><span class="co">#&gt; 2                NA              NA                 NA          NA</span></span>
<span><span class="co">#&gt; 3                NA              NA                 NA          NA</span></span>
<span><span class="co">#&gt;   ndf_chisq_heter pvalue_heter chisq_residual ndf_residual pvalue_residual</span></span>
<span><span class="co">#&gt; 1              NA           NA             NA           NA              NA</span></span>
<span><span class="co">#&gt; 2              NA           NA             NA           NA              NA</span></span>
<span><span class="co">#&gt; 3              NA           NA             NA           NA              NA</span></span>
<span><span class="co">#&gt;           comments</span></span>
<span><span class="co">#&gt; 1 SmallcohortCount</span></span>
<span><span class="co">#&gt; 2 SmallcohortCount</span></span>
<span><span class="co">#&gt; 3 SmallcohortCount</span></span></code></pre></div>
<p>In <code>MR_mega_prep</code> function, the arguments are described in
details here:</p>
<ul>
<li><p><code>sum_stat_pop</code> is a list of all GWAS file data which
would be inputted in the environment-adjuste meta-regression model or
the original meta-regression model.</p></li>
<li><p><code>n_cohort</code> refers to the number of cohorts.</p></li>
<li><p><code>usefor</code> is the option of meta analysis methods which
can be “mr-mega” or “env-mr-mega”. If the option is “mr-mega”, the
limitation of the number of axes of genetic variation (<span style="color: #0000FF;">T</span>) should be <span style="color: #0000FF;">T&lt;K-2</span>. If the option is “env-mr-mega”,
the limitation of the number of axes of genetic variation (<span style="color: #0000FF;">T</span>) and the number of environmental
covariates (<span style="color: #0000FF;">S</span>) should be <span style="color: #0000FF;">T+S&lt;K-2</span>.</p></li>
<li><p><code>qt</code> is logical vector of length 1, which means
whether the trait is quantitative (TRUE) or not (FALSE). The current
package is only applied for quantitative traits. Default is
TRUE.</p></li>
<li><p><code>pcCount</code> is the number of axes of genetic variation
(<span style="color: #0000FF;">T</span>).</p></li>
<li><p><code>calpc</code> is logical vector of length 1, which means
whether need to calculate PCs. If PCs have been obtained,
<code>calpc=FALSE</code>.</p></li>
<li><p><code>envCount</code> is the number of environmental covariates
(<span style="color: #0000FF;">S</span>).</p></li>
</ul>
<p>In the procedure of pre-processing the GWAS data, we will filter out
the genetic variants of which the frequencies less than
<code>pcCount</code>+2 for ‘mr-mega’ and <code>pcCount+envCount</code>+2
for ‘env-mr-mega’. These genetic variants will not be inputted into the
analysis of meta-regression models and saved as <code>sel_gene</code>
object in the output of <code>MR_mega_prep</code> function. As
<strong>GWAS data format</strong> section mentioned, the frequencies of
the four genetic variants (<code>rs190897515</code>,
<code>rs577430535</code>,<code>rs185581532</code>,<code>rs572565163</code>)
existing across the 7 populations are 1,1,5,4. For
<code>usefor="mr-mega"</code>, meta-analysis would exclude three genetic
variants (<code>rs190897515</code>,
<code>rs577430535</code>,<code>rs572565163</code>) due to <span style="color: #0000FF;">T&lt;K-2</span>. For
<code>usefor="env-mr-mega"</code>, meta-analysis would exclude three
genetic variants (<code>rs190897515</code>,
<code>rs577430535</code>,<code>rs185581532,</code>rs572565163`) due to
<span style="color: #0000FF;">T+S&lt;K-2</span>.</p>
<p>In addition to <code>sel_gene</code> object, the output of
<code>MR_mega_prep</code> function contain other objects: BETA values
(<code>BETA</code>), the inverse of variance (<code>invse2_pop</code>),
the information of each genetic variant (<code>marker_inf_pop</code>),
the frequencies of remained genetic variants across the 7 cohorts
(<code>cohort_count_filt</code>). If <code>calpc=TRUE</code>, axes of
genetic variation would be calculated based on the inputted
<code>sum_stat_pop</code> and the resulting output would contain
<code>pcs</code> object.</p>
<p>In this simulation, we have calculated 2 PCs which would be used in
the original meta-regression model and environment-adjusted
meta-regression model. Besides, for environment-adjusted meta-regression
model accounting for environmental exposures varied between cohorts, we
introduce individual-level smoking status as a environmental covariate.
For each environmental covariate measured in all cohorts, we construct a
cohort-level environmental covariate by taking the mean or proportion of
each covariate within each cohort. Here we give the smoking proportions
across the 7 populations.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">smoke.env</span><span class="op">)</span></span>
<span><span class="co">#&gt;   acb asw esn gwd lwk msl yri</span></span>
<span><span class="co">#&gt; 1   1   0   0   1   1   0   0</span></span>
<span><span class="co">#&gt; 2   1   0   1   1   0   1   1</span></span>
<span><span class="co">#&gt; 3   1   0   0   1   0   0   1</span></span>
<span><span class="co">#&gt; 4   0   0   0   1   1   0   1</span></span>
<span><span class="co">#&gt; 5   1   1   0   0   1   1   1</span></span>
<span><span class="co">#&gt; 6   0   1   0   1   0   1   0</span></span>
<span><span class="va">smoke.env</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">smoke.env</span>,<span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>,<span class="fl">4</span><span class="op">)</span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">smoke.env</span></span>
<span><span class="co">#&gt;    acb    asw    esn    gwd    lwk    msl    yri </span></span>
<span><span class="co">#&gt; 0.7922 0.1750 0.2130 0.6860 0.5510 0.2015 0.7850</span></span></code></pre></div>
<p>To prepare performing meta-analysis by the original MR-MEGA approach
(<code>MR_mega_lrw</code>) and the environment-adjusted MR-MEGA approach
(<code>MR_mega_env_lrw</code>), we need to add axes of genetic variants
(PCs) into <code>gwas_input</code> object. Especially for the
environment-adjusted MR-MEGA approach, cohort-level environmental
covariates would be added into <code>gwas_input</code> object as well.
If we set ‘calpc=TRUE’ in <code>MR_mega_prep</code> function, the
resulting output of <code>MR_mega_prep</code> function would comprise
<code>pcs</code> object.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gwas_input</span><span class="op">$</span><span class="va">pcs</span><span class="op">=</span><span class="va">pcs</span>;</span>
<span><span class="va">gwas_input</span><span class="op">$</span><span class="va">env</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">smoke.env</span>,<span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>,<span class="fl">4</span><span class="op">)</span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><span style="color: #0000FF;">Note</span>:</p>
<ul>
<li>For 14 sex-stratified cohorts, <code>gwas_input$env</code> would be
shown as a dataframe, consisting of a vector for smoking proportions
across cohorts and a vector for sex (female cohorts=1/male
cohorts=0).</li>
</ul>
</div>
<div class="section level3">
<h3 id="gwas-data-analysis">GWAS data analysis<a class="anchor" aria-label="anchor" href="#gwas-data-analysis"></a>
</h3>
<p>After constructing a list of the inputted GWAS data
<code>gwas_input</code> obtained from <code>MR_mega_prep</code>
function, we then use <code>MR_mega_lrw</code> function and
<code>MR_mega_env_lrw</code> function to perform meta analysis
respectively.</p>
<p><span style="color: #0000FF;">Note</span>: In the original
meta-regression model, it is not required that <code>gwas_input</code>
must contain <code>env</code> object.</p>
<p><strong>MR-mega</strong></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gwas_input</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "beta_pop"          "invse2_pop"        "marker_inf_pop"   </span></span>
<span><span class="co">#&gt; [4] "cohort_count_filt" "sel_gene"          "pcs"              </span></span>
<span><span class="co">#&gt; [7] "env"</span></span>
<span><span class="fu">tictoc</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tictoc/man/tic.html" class="external-link">tic</a></span><span class="op">(</span><span class="st">"Perform MR-MEGA approach"</span><span class="op">)</span></span>
<span><span class="va">out</span><span class="op">=</span><span class="fu"><a href="../reference/MR_mega_lrw.html">MR_mega_lrw</a></span><span class="op">(</span><span class="va">gwas_input</span>,qt<span class="op">=</span><span class="cn">TRUE</span>,ncores<span class="op">=</span><span class="fl">2</span>,pcCount<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Linear regression model for each genetic varaiant"</span></span>
<span><span class="fu">tictoc</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tictoc/man/tic.html" class="external-link">toc</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Perform MR-MEGA approach: 0.296 sec elapsed</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">out</span>,<span class="fl">4</span><span class="op">)</span></span>
<span><span class="co">#&gt;    MARKERNAME CHROMOSOME EA NEA     N   EAF_avg      dir        beta0</span></span>
<span><span class="co">#&gt; 1  rs10429359          8  C   G 28000 0.3426143 ?--+++-+  0.003316407</span></span>
<span><span class="co">#&gt; 2 rs115995911         10  T   C 28000 0.1558014 ?----++- -0.008618561</span></span>
<span><span class="co">#&gt; 3   rs2419893         13  A   G 28000 0.1334857 ?+--++-- -0.001259743</span></span>
<span><span class="co">#&gt; 4 rs201029612          1  C   T 28000 0.4912714 ?+-+---- -0.028488694</span></span>
<span><span class="co">#&gt;           se0       beta1       se1        beta2       se2 chisq_association</span></span>
<span><span class="co">#&gt; 1 0.004638613  0.22445318 0.2003457 -0.430731581 0.2320187        6.94648260</span></span>
<span><span class="co">#&gt; 2 0.007113917 -0.08141943 0.2774925 -0.014201005 0.3972281        2.90025500</span></span>
<span><span class="co">#&gt; 3 0.004648444 -0.04990717 0.2088015 -0.003257743 0.2283215        0.09731795</span></span>
<span><span class="co">#&gt; 4 0.028640912 -0.18168184 1.3195394  0.657007859 1.4035409       44.24654275</span></span>
<span><span class="co">#&gt;   ndf_association pvalue_association chisq_heter ndf_heter pvalue_heter</span></span>
<span><span class="co">#&gt; 1               3       7.362326e-02  6.27891298         2   0.04330633</span></span>
<span><span class="co">#&gt; 2               3       4.072609e-01  0.14426046         2   0.93040972</span></span>
<span><span class="co">#&gt; 3               3       9.921573e-01  0.03962839         2   0.98038082</span></span>
<span><span class="co">#&gt; 4               3       1.337727e-09  7.68822286         2   0.02140541</span></span>
<span><span class="co">#&gt;   chisq_residual ndf_residual pvalue_residual Count_cohorts</span></span>
<span><span class="co">#&gt; 1       5.352937            4    2.529621e-01             7</span></span>
<span><span class="co">#&gt; 2       6.702019            4    1.524983e-01             7</span></span>
<span><span class="co">#&gt; 3       2.725580            4    6.047459e-01             7</span></span>
<span><span class="co">#&gt; 4     129.937830            4    4.015115e-27             7</span></span></code></pre></div>
<p>For the original MR-MEGA approach, results contain following
columns:</p>
<ul>
<li><p><code>MARKERNAME</code>: unique marker identification across all
cohorts.</p></li>
<li><p><code>CHROMOSOME</code>: chrmosome of marker.</p></li>
<li><p><code>POSITION</code>: physical position in chromosome of
marker.</p></li>
<li><p><code>EA</code>: effective allele.</p></li>
<li><p><code>NEA</code>: non-effective allele.</p></li>
<li><p><code>N</code>: sample size.</p></li>
<li><p><code>EAF_avg</code>: the weighted EAF.</p></li>
<li><p><code>dir</code>: the direction of allelic effects (BETA) across
all cohorts. In the cohort, the allelic effect is is denoted as follows:
“+”, if&gt;0; “-”, if &lt;/=0; NA, if missing.</p></li>
<li><p><code>beta0</code>: effect of the intercept of
meta-regression.</p></li>
<li><p><code>se0</code>: stderr of the effect of the intercept of
meta-regression.</p></li>
<li><p><code>beta{i}</code>: effect of the i-th PC of
meta-regression.</p></li>
<li><p><code>se{i}</code>: stderr of the effect of the i-th PC of
meta-regression.</p></li>
<li><p><code>chisq_association</code>, <code>ndf_association</code> and
<code>pvalue_association</code>: test of the null hypothesis of no
association of each variant. <code>chisq_</code>, <code>ndf_</code> and
<code>pvalue_</code> refer to chisq value, number of degrees of freedom
and p-values of the corresponding test.</p></li>
<li><p><code>chisq_heter</code>, <code>ndf_heter</code> and
<code>pvalue_heter</code>: test of heterogeneity in allelic effects. For
the orignial MR-MEGA approach, it is assumed that the heterogeneity in
allelic effects might be correlated with the ancestral effect
only.</p></li>
<li><p><code>chisq_residual</code>, <code>ndf_residual</code> and
<code>pvalue_residual</code>: test of residual heterogeneity.</p></li>
<li><p><code>Count_cohorts</code>: the frequencies of genetic variants
present in all cohorts.</p></li>
</ul>
<p><strong>Environment-adjusted MR-mega</strong></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">tictoc</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tictoc/man/tic.html" class="external-link">tic</a></span><span class="op">(</span><span class="st">"Perform env-MR-MEGA approach"</span><span class="op">)</span></span>
<span><span class="va">out.env</span><span class="op">=</span><span class="fu"><a href="../reference/MR_mega_env_lrw.html">MR_mega_env_lrw</a></span><span class="op">(</span><span class="va">gwas_input</span>,qt<span class="op">=</span><span class="cn">TRUE</span>,ncores<span class="op">=</span><span class="fl">2</span>,pcCount<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Linear regression model for each genetic varaiant"</span></span>
<span><span class="fu">tictoc</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tictoc/man/tic.html" class="external-link">toc</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Perform env-MR-MEGA approach: 0.334 sec elapsed</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">out.env</span>,<span class="fl">4</span><span class="op">)</span></span>
<span><span class="co">#&gt;    MARKERNAME CHROMOSOME EA NEA     N   EAF_avg      dir        beta0</span></span>
<span><span class="co">#&gt; 1  rs10429359          8  C   G 28000 0.3426143 ?--+++-+ -0.011451926</span></span>
<span><span class="co">#&gt; 2 rs115995911         10  T   C 28000 0.1558014 ?----++- -0.001497157</span></span>
<span><span class="co">#&gt; 3   rs2419893         13  A   G 28000 0.1334857 ?+--++-- -0.008231415</span></span>
<span><span class="co">#&gt; 4 rs201029612          1  C   T 28000 0.4912714 ?+-+----  0.010190606</span></span>
<span><span class="co">#&gt;           se0       beta1        se1       beta2       se2       beta3</span></span>
<span><span class="co">#&gt; 1 0.006713941  0.02992510 0.01209679  0.15795892 0.1353795 -0.37156698</span></span>
<span><span class="co">#&gt; 2 0.016379065 -0.01489393 0.03001156 -0.03544404 0.3216560 -0.04460317</span></span>
<span><span class="co">#&gt; 3 0.011129979  0.01386003 0.01979218 -0.08123241 0.2279574  0.02450359</span></span>
<span><span class="co">#&gt; 4 0.070106808 -0.07477172 0.12140240 -0.04802027 1.4519218  0.54035478</span></span>
<span><span class="co">#&gt;         se3 chisq_association ndf_association pvalue_association chisq_heter</span></span>
<span><span class="co">#&gt; 1 0.1555107        10.5385290               4       3.227038e-02   9.8709594</span></span>
<span><span class="co">#&gt; 2 0.4451722         3.4087188               4       4.918929e-01   0.6527242</span></span>
<span><span class="co">#&gt; 3 0.2476158         0.4802537               4       9.753946e-01   0.4225641</span></span>
<span><span class="co">#&gt; 4 1.5387035        58.8321606               4       5.103427e-12  22.2738408</span></span>
<span><span class="co">#&gt;   ndf_heter pvalue_heter chisq_residual ndf_residual pvalue_residual  chisq_env</span></span>
<span><span class="co">#&gt; 1         3 1.969549e-02       1.760890            3    6.234824e-01  3.5920464</span></span>
<span><span class="co">#&gt; 2         3 8.842632e-01       6.193555            3    1.025638e-01  0.5084638</span></span>
<span><span class="co">#&gt; 3         3 9.355428e-01       2.342644            3    5.044007e-01  0.3829358</span></span>
<span><span class="co">#&gt; 4         3 5.720874e-05     115.352212            3    7.731354e-25 14.5856179</span></span>
<span><span class="co">#&gt;   ndf_env   pvalue_env   chisq_PC ndf_PC  pvalue_PC Count_cohorts</span></span>
<span><span class="co">#&gt; 1       1 0.0580567103 4.04708318      2 0.13218648             7</span></span>
<span><span class="co">#&gt; 2       1 0.4758046426 0.04296995      2 0.97874418             7</span></span>
<span><span class="co">#&gt; 3       1 0.5360362428 0.10221375      2 0.95017712             7</span></span>
<span><span class="co">#&gt; 4       1 0.0001339328 4.76185690      2 0.09246469             7</span></span></code></pre></div>
<p>For the environment-adjusted MR-MEGA approach, results contain tests
which are different from that in the original MR-MEGA approach. Details
are shown as follows:</p>
<ul>
<li><p><code>chisq_heter</code>, <code>ndf_heter</code> and
<code>pvalue_heter</code>: test of heterogeneity in allelic and
environmental effects.</p></li>
<li><p><code>chisq_env</code>, <code>ndf_env</code> and
<code>pvalue_env</code>: test of heterogeneity in the environmental
effect only.</p></li>
<li><p><code>chisq_PC</code>, <code>ndf_PC</code> and
<code>pvalue_PC</code>: test of heterogeneity in the allelic effect
only.</p></li>
</ul>
</div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="reference">Reference<a class="anchor" aria-label="anchor" href="#reference"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-10002015global" class="csl-entry">
Consortium, 1000 Genomes Project et al. 2015. <span>“A Global Reference
for Human Genetic Variation.”</span> <em>Nature</em> 526 (7571): 68.
</div>
<div id="ref-magi2017trans" class="csl-entry">
Mägi, Reedik, Momoko Horikoshi, Tamar Sofer, Anubha Mahajan, Hidetoshi
Kitajima, Nora Franceschini, Mark I McCarthy, T2D-GENES Consortium
COGENT-Kidney Consortium, and Andrew P Morris. 2017. <span>“Trans-Ethnic
Meta-Regression of Genome-Wide Association Studies Accounting for
Ancestry Increases Power for Discovery and Improves Fine-Mapping
Resolution.”</span> <em>Human Molecular Genetics</em> 26 (18): 3639–50.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Siru Wang, Jennifer Asimit, Andrew P Morris.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
